plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs

```diff
@@ -0,0 +1,202 @@
+foldAst :: (Monad m) => HieAST t -> Tokenizer m ()
+foldAst ast = if null (nodeChildren ast)
+  then liftMaybeM (visitLeafIds ast)
+  else mapM_ foldAst $ nodeChildren ast
+
+visitLeafIds :: HieAST t -> Tokenizer Maybe ()
+visitLeafIds leaf = liftMaybeM $ do
+  let span = nodeSpan leaf
+  (ran, token) <- focusTokenAt leaf
+  -- if `focusTokenAt` succeed, we can safely assume we have shift the cursor correctly
+  -- we do not need to recover the cursor state, even if the following computation failed
+  liftMaybeM $ do
+    -- only handle the leaf node with single column token
+    guard $ srcSpanStartLine span == srcSpanEndLine span
+    splitResult <-  lift $ splitRangeByText token ran
+    modify $ \s -> s {currentRange = ran, currentRangeContext = splitResult}
+    mapM_ combineNodeIds $ Map.filterWithKey (\k _ -> k == SourceInfo) $ getSourcedNodeInfo $ sourcedNodeInfo leaf
+  where
+    combineNodeIds :: (Monad m) => NodeInfo a -> Tokenizer m ()
+    combineNodeIds (NodeInfo _ _ bd) = mapM_ getIdentifier (M.keys bd)
+    getIdentifier :: (Monad m) => Identifier -> Tokenizer m ()
+    getIdentifier idt = liftMaybeM $ do
+      ran <- gets currentRange
+      case idt of
+        Left _moduleName -> addRangeIdSetMap ran idt
+        Right name -> do
+          ranSplit <- gets currentRangeContext
+          occStr <- lift $ case (occNameString . nameOccName) name of
+            -- the generated selector name with {-# LANGUAGE DuplicateRecordFields #-}
+            '$' : 's' : 'e' : 'l' : ':' : xs -> Just $ takeWhile (/= ':') xs
+            ['$']                            -> Just "$"
+            -- other generated names that should not be visible
+            '$' : _                          -> Nothing
+            ns                               -> Just ns
+          case ranSplit of
+            (NoSplit (tk, r)) -> do
+              guard $ T.unpack tk == occStr
+              addRangeIdSetMap r idt
+            (Split (tk, r1, r2)) -> do
+              guard $ T.unpack tk == occStr
+              addRangeIdSetMap r1 (Left $ mkModuleName "")
+              addRangeIdSetMap r2 idt
+
+focusTokenAt ::
+  -- | leaf node we want to focus on
+  HieAST a ->
+  -- | (token, remains)
+  Tokenizer Maybe (Range, Text)
+focusTokenAt leaf = do
+  PTokenState{cursor, rope, columnsInUtf16} <- get
+  let span = nodeSpan leaf
+  let (startPos, endPos) = srcSpanCharPositions span
+  startOff <- lift $ startPos `sub` cursor
+  tokenOff <- lift $ endPos `sub` startPos
+  (gap, startRope) <- lift $ charSplitAtPositionMaybe startOff rope
+  (token, remains) <- lift $ charSplitAtPositionMaybe tokenOff startRope
+  let ncs = newColumn columnsInUtf16 gap
+  let nce = newColumn ncs token
+  -- compute the new range for utf16, tuning the columns is enough
+  let ran = codePointRangeToRangeWith ncs nce $ realSrcSpanToCodePointRange span
+  modify $ \s -> s {columnsInUtf16 = nce, rope = remains, cursor = endPos}
```

Comment:
This stretch of code is a bit hard to follow. Perhaps it could use more comments explaining what is going on?

Here is my attempt:

We start with a `rope` representing the document remaining to be processed. `cursor` is a position (utf8?) relative to the start of the document, and is relative to the entire document, not the fragment we have in the `rope`. We want to focus on the position given by the `nodeSpan` of the `HieAST`, and return the token `(Range, Text)` at that span, advancing the `cursor` and the `rope` past the span. `nodeSpan`(utf8?) is also relative to the start of the entire document and not the fragment. `columnsInUtf16` is the number of columns in the utf8 encoding of the current line.

To obtain offsets corresponding to the `nodeSpan` in the fragment, we must subtract the `cursor` from the start position of the `nodeSpan`. This gives us a (utf8?) offset into `rope`. We subtract the start position from the endposition to obtain the (utf8?) length of the token. We can then advance the `cursor` and the `rope` past the token, obtaining the (utf8?) text of the current token in the process.

To obtain the utf16 range, we must advance the current columns first past the initial segment of the rope that we skip past (using its utf16 length, and accounting for newlines by only using the length from the last newline to the end of the segment), and then again past the current token (similarly). The first skip gives us the start column in utf16 bytes and the second skip gives us the end column, also in utf16 bytes. We finally update the `columnsInUtf16` using the end column.

Reply:
yes, you are right, some renaming and doc are added now to make the code more readable

Reply:
Some comments are added to PTokenState
and Some added to this stretch of the code.